### PR TITLE
feat: 0.45.4 stride 4

### PR DESCRIPTION
**Pull request by riley-stride**
## Summary
Redirect staking rewards away from any delegators to the blacklisted validator, to the remaining validators

## Details
In `allocation.go`:
- set the blocked addr on line 14
- calculate the "power" (vote share) of this validator on lines 28-34. note that defaultpowerreduction is just the conversion rate between ustrd => strd afaiu! (would be nice to get some confirmation on this though :))
- On line 105, deduct this validator's power from the total power, so that the other validator's rewards are upscaled proportionally. This redirects the blocked validator's rewards to the others
- On lines 107-113, set the blocked validator's outstanding rewards to 0

Note that we allow the blocked validator to accrue proposer rewards (lines 61-79), which are very small.

## Test
Follow the instructions in this [pull request](https://github.com/Stride-Labs/stride/pull/128/)

### CODE REVIEW

File: x/distribution/keeper/allocation.go [hunk 1]
```diff
@@ -90,8 +102,15 @@ func (k Keeper) AllocateTokens(
 
 		// TODO consider microslashing for missing votes.
 		// ref https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701
-		powerFraction := sdk.NewDec(vote.Validator.Power).QuoTruncate(sdk.NewDec(totalPreviousPower))
+		powerFraction := sdk.NewDec(vote.Validator.Power).QuoTruncate(sdk.NewDec(totalPreviousPower - blacklisted_val_power))
```
Comment (asalzmann) [inline x/distribution/keeper/allocation.go:105]: nit: pull `totalPreviousPower - blacklisted_val_power` out of the for loop for clarity

File: x/distribution/keeper/allocation.go [hunk 2]
```diff
@@ -90,8 +102,15 @@ func (k Keeper) AllocateTokens(
...
 		reward := feesCollected.MulDecTruncate(voteMultiplier).MulDecTruncate(powerFraction)
+
+		valAddr := validator.GetOperator().String()
+		if valAddr == BLACKLISTED_VAL_ADDR {
```
Comment (asalzmann) [inline x/distribution/keeper/allocation.go:109]:
```suggestion
		if valAddr == BLACKLISTED_VAL_ADDR {
		    continue
```
Can we add a line like this to the top of the loop instead?

**Review by asalzmann — COMMENTED** [review]
Looks good overall! A few comments

1. Wdyt of generalizing this to handle multiple validators?
2. Is there any way we can parameterize the validators?
3. Can we get a review from someone that's worked on the staking module?

**Comment by riley-stride** [comment]
> Looks good overall! A few comments
> 
> 1. Wdyt of generalizing this to handle multiple validators?
> 2. Is there any way we can parameterize the validators?
> 3. Can we get a review from someone that's worked on the staking module?

Good suggestions, all addressed in latest commit. Val list is a param in distr module now, updateable by governance.

**Pull request closed without merging**